File: hayeah/src/hayeah/fzfmatch.py
"""Non-interactive fuzzy path matcher inspired by fzf extended-search syntax.

Deterministic boolean filter — no scoring, no ranking. Paths are matched
case-insensitively with forward-slash normalization.

Term syntax::

    [!]['][^]<text>[$][']

Expression composition::

    term1 term2       — implicit AND (all terms must match)
    expr | expr       — compound AND (intersection, lower precedence)
    expr ; expr       — union OR (higher precedence)
"""

from __future__ import annotations

import unicodedata
from dataclasses import dataclass, field


class MatchError(ValueError):
    """Raised when a pattern is malformed."""


# ---------------------------------------------------------------------------
# Word-boundary helpers
# ---------------------------------------------------------------------------

def _is_word_char(ch: str) -> bool:
    """Letter or digit. Underscore is NOT a word char (unlike regex \\w)."""
    return ch.isalpha() or ch.isdigit()


def _has_word_boundary(s: str, idx: int, size: int) -> bool:
    left_ok = idx == 0 or not _is_word_char(s[idx - 1])
    right_ok = idx + size == len(s) or not _is_word_char(s[idx + size])
    return left_ok and right_ok


def contains_word_exact(s: str, needle: str) -> bool:
    """True if *needle* appears in *s* bounded on both sides."""
    if not needle:
        return False
    start = 0
    while start <= len(s) - len(needle):
        idx = s.find(needle, start)
        if idx < 0:
            break
        if _has_word_boundary(s, idx, len(needle)):
            return True
        start = idx + 1
    return False


def contains_word_prefix(s: str, needle: str) -> bool:
    """True if *needle* appears in *s* with a word boundary on the left."""
    if not needle:
        return False
    start = 0
    while start <= len(s) - len(needle):
        idx = s.find(needle, start)
        if idx < 0:
            break
        if idx == 0 or not _is_word_char(s[idx - 1]):
            return True
        start = idx + 1
    return False


# ---------------------------------------------------------------------------
# Term
# ---------------------------------------------------------------------------

@dataclass
class _Term:
    raw: str
    text: str = ""
    anchor_head: bool = False
    anchor_tail: bool = False
    word_prefix: bool = False
    word_exact: bool = False
    neg: bool = False


def _parse_term(raw: str) -> _Term:
    t = _Term(raw=raw)
    p = raw

    # negation
    if p.startswith("!"):
        t.neg = True
        p = p[1:]
        if not p:
            raise MatchError(f"empty term after negation in {raw!r}")

    # word-boundary quotes
    if p.startswith("'"):
        p = p[1:]
        if not p:
            raise MatchError(f"empty term after leading quote in {raw!r}")
        if p.endswith("'"):
            t.word_exact = True
            p = p[:-1]
            if not p:
                raise MatchError(f"empty term in {raw!r}")
        else:
            t.word_prefix = True

    # ./ as ^ anchor
    if p.startswith("./"):
        t.anchor_head = True
        p = p[2:]

    # ^ / $ anchors
    if p.startswith("^"):
        t.anchor_head = True
        p = p[1:]
    if p.endswith("$"):
        t.anchor_tail = True
        p = p[:-1]

    if not p:
        raise MatchError(f"empty term after stripping modifiers in {raw!r}")

    t.text = p.lower().replace("\\", "/")
    return t


def _term_matches(t: _Term, path: str) -> bool:
    # exact path fast path
    if t.anchor_head and t.anchor_tail and not (t.word_exact or t.word_prefix):
        return path == t.text

    sub = path
    if t.anchor_head:
        if not path.startswith(t.text):
            return False
        sub = path[: len(t.text)]
    if t.anchor_tail:
        if not path.endswith(t.text):
            return False
        sub = path[len(path) - len(t.text) :]

    if t.word_exact:
        return contains_word_exact(sub, t.text)
    elif t.word_prefix:
        return contains_word_prefix(sub, t.text)
    else:
        return t.text in sub


# ---------------------------------------------------------------------------
# Matcher protocol
# ---------------------------------------------------------------------------

class Matcher:
    """Base class for matchers."""

    def match(self, paths: list[str]) -> list[str]:
        raise NotImplementedError


@dataclass
class FuzzyMatcher(Matcher):
    """Matches paths against space-separated terms (implicit AND)."""

    pattern: str
    _terms: list[_Term] = field(default_factory=list, repr=False)

    def match(self, paths: list[str]) -> list[str]:
        if not self._terms:
            return list(paths)

        out: list[str] = []
        for path in paths:
            normal = path.lower().replace("\\", "/")
            ok = True
            for term in self._terms:
                m = _term_matches(term, normal)
                if term.neg:
                    m = not m
                if not m:
                    ok = False
                    break
            if ok:
                out.append(path)
        return out


@dataclass
class CompoundMatcher(Matcher):
    """AND — chains matchers sequentially (intersection)."""

    matchers: list[Matcher]

    def match(self, paths: list[str]) -> list[str]:
        current = paths
        for m in self.matchers:
            current = m.match(current)
        return current


@dataclass
class UnionMatcher(Matcher):
    """OR — merges results from all matchers (deduplicated, preserves first-seen order)."""

    matchers: list[Matcher]

    def match(self, paths: list[str]) -> list[str]:
        seen: set[str] = set()
        out: list[str] = []
        for m in self.matchers:
            for p in m.match(paths):
                if p not in seen:
                    seen.add(p)
                    out.append(p)
        return out


# ---------------------------------------------------------------------------
# Parser
# ---------------------------------------------------------------------------

def _new_fuzzy_matcher(pattern: str) -> FuzzyMatcher:
    pattern = pattern.strip()
    if not pattern:
        return FuzzyMatcher(pattern=pattern)
    terms = [_parse_term(tok) for tok in pattern.split()]
    return FuzzyMatcher(pattern=pattern, _terms=terms)


def parse_matcher(pattern: str) -> Matcher:
    """Parse a pattern string into a Matcher.

    Operators:
        ``|`` — compound AND (lower precedence, splits first)
        ``;`` — union OR (higher precedence, binds tighter)
    """
    pattern = pattern.strip()

    # pipe = AND (lower precedence — checked first)
    if "|" in pattern:
        parts = pattern.split("|")
        matchers = [parse_matcher(p) for p in parts if p.strip()]
        if not matchers:
            raise MatchError("compound pattern contains no valid patterns")
        if len(matchers) == 1:
            return matchers[0]
        return CompoundMatcher(matchers=matchers)

    # semicolon = OR (higher precedence)
    if ";" in pattern:
        parts = pattern.split(";")
        matchers = [parse_matcher(p) for p in parts if p.strip()]
        if not matchers:
            raise MatchError("union pattern contains no valid patterns")
        if len(matchers) == 1:
            return matchers[0]
        return UnionMatcher(matchers=matchers)

    return _new_fuzzy_matcher(pattern)

File: libs/hayeah-go/fzfmatch/fzfmatch.go
// Package fzfmatch provides a non-interactive fuzzy path matcher inspired
// by fzf extended-search syntax. Deterministic boolean filter — no scoring,
// no ranking. Paths are matched case-insensitively with forward-slash
// normalization.
package fzfmatch

import (
	"errors"
	"fmt"
	"strings"
	"unicode"
)

// MatchError is returned when a pattern is malformed.
var MatchError = errors.New("match error")

// ---------------------------------------------------------------------------
// Word-boundary helpers
// ---------------------------------------------------------------------------

func isWordChar(ch rune) bool {
	return unicode.IsLetter(ch) || unicode.IsDigit(ch)
}

func hasWordBoundary(s string, idx, size int) bool {
	leftOK := idx == 0 || !isWordChar(rune(s[idx-1]))
	rightOK := idx+size == len(s) || !isWordChar(rune(s[idx+size]))
	return leftOK && rightOK
}

// ContainsWordExact returns true if needle appears in s bounded on both sides.
func ContainsWordExact(s, needle string) bool {
	if needle == "" {
		return false
	}
	start := 0
	for start <= len(s)-len(needle) {
		idx := strings.Index(s[start:], needle)
		if idx < 0 {
			break
		}
		idx += start
		if hasWordBoundary(s, idx, len(needle)) {
			return true
		}
		start = idx + 1
	}
	return false
}

// ContainsWordPrefix returns true if needle appears in s with a word boundary on the left.
func ContainsWordPrefix(s, needle string) bool {
	if needle == "" {
		return false
	}
	start := 0
	for start <= len(s)-len(needle) {
		idx := strings.Index(s[start:], needle)
		if idx < 0 {
			break
		}
		idx += start
		if idx == 0 || !isWordChar(rune(s[idx-1])) {
			return true
		}
		start = idx + 1
	}
	return false
}

// ---------------------------------------------------------------------------
// Term
// ---------------------------------------------------------------------------

type term struct {
	raw        string
	text       string
	anchorHead bool
	anchorTail bool
	wordPrefix bool
	wordExact  bool
	neg        bool
}

func parseTerm(raw string) (term, error) {
	t := term{raw: raw}
	p := raw

	// negation
	if strings.HasPrefix(p, "!") {
		t.neg = true
		p = p[1:]
		if p == "" {
			return t, fmt.Errorf("%w: empty term after negation in %q", MatchError, raw)
		}
	}

	// word-boundary quotes
	if strings.HasPrefix(p, "'") {
		p = p[1:]
		if p == "" {
			return t, fmt.Errorf("%w: empty term after leading quote in %q", MatchError, raw)
		}
		if strings.HasSuffix(p, "'") {
			t.wordExact = true
			p = p[:len(p)-1]
			if p == "" {
				return t, fmt.Errorf("%w: empty term in %q", MatchError, raw)
			}
		} else {
			t.wordPrefix = true
		}
	}

	// ./ as ^ anchor
	if strings.HasPrefix(p, "./") {
		t.anchorHead = true
		p = p[2:]
	}

	// ^ / $ anchors
	if strings.HasPrefix(p, "^") {
		t.anchorHead = true
		p = p[1:]
	}
	if strings.HasSuffix(p, "$") {
		t.anchorTail = true
		p = p[:len(p)-1]
	}

	if p == "" {
		return t, fmt.Errorf("%w: empty term after stripping modifiers in %q", MatchError, raw)
	}

	t.text = strings.ToLower(strings.ReplaceAll(p, "\\", "/"))
	return t, nil
}

func termMatches(t term, path string) bool {
	// exact path fast path
	if t.anchorHead && t.anchorTail && !t.wordExact && !t.wordPrefix {
		return path == t.text
	}

	sub := path
	if t.anchorHead {
		if !strings.HasPrefix(path, t.text) {
			return false
		}
		sub = path[:len(t.text)]
	}
	if t.anchorTail {
		if !strings.HasSuffix(path, t.text) {
			return false
		}
		sub = path[len(path)-len(t.text):]
	}

	if t.wordExact {
		return ContainsWordExact(sub, t.text)
	} else if t.wordPrefix {
		return ContainsWordPrefix(sub, t.text)
	}
	return strings.Contains(sub, t.text)
}

// ---------------------------------------------------------------------------
// Matcher
// ---------------------------------------------------------------------------

// Matcher filters a list of paths.
type Matcher interface {
	Match(paths []string) []string
}

// FuzzyMatcher matches paths against space-separated terms (implicit AND).
type FuzzyMatcher struct {
	Pattern string
	terms   []term
}

func (m *FuzzyMatcher) Match(paths []string) []string {
	if len(m.terms) == 0 {
		out := make([]string, len(paths))
		copy(out, paths)
		return out
	}

	var out []string
	for _, path := range paths {
		normal := strings.ToLower(strings.ReplaceAll(path, "\\", "/"))
		ok := true
		for _, t := range m.terms {
			matched := termMatches(t, normal)
			if t.neg {
				matched = !matched
			}
			if !matched {
				ok = false
				break
			}
		}
		if ok {
			out = append(out, path)
		}
	}
	return out
}

// CompoundMatcher chains matchers sequentially (AND / intersection).
type CompoundMatcher struct {
	Matchers []Matcher
}

func (m *CompoundMatcher) Match(paths []string) []string {
	current := paths
	for _, sub := range m.Matchers {
		current = sub.Match(current)
	}
	return current
}

// UnionMatcher merges results from all matchers (OR, deduplicated, first-seen order).
type UnionMatcher struct {
	Matchers []Matcher
}

func (m *UnionMatcher) Match(paths []string) []string {
	seen := make(map[string]bool)
	var out []string
	for _, sub := range m.Matchers {
		for _, p := range sub.Match(paths) {
			if !seen[p] {
				seen[p] = true
				out = append(out, p)
			}
		}
	}
	return out
}

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

func newFuzzyMatcher(pattern string) (*FuzzyMatcher, error) {
	pattern = strings.TrimSpace(pattern)
	if pattern == "" {
		return &FuzzyMatcher{Pattern: pattern}, nil
	}
	tokens := strings.Fields(pattern)
	terms := make([]term, 0, len(tokens))
	for _, tok := range tokens {
		t, err := parseTerm(tok)
		if err != nil {
			return nil, err
		}
		terms = append(terms, t)
	}
	return &FuzzyMatcher{Pattern: pattern, terms: terms}, nil
}

// ParseMatcher parses a pattern string into a Matcher.
//
// Operators:
//
//	| — compound AND (lower precedence, splits first)
//	; — union OR (higher precedence, binds tighter)
func ParseMatcher(pattern string) (Matcher, error) {
	pattern = strings.TrimSpace(pattern)

	// pipe = AND (lower precedence — checked first)
	if strings.Contains(pattern, "|") {
		parts := strings.Split(pattern, "|")
		var matchers []Matcher
		for _, p := range parts {
			if strings.TrimSpace(p) == "" {
				continue
			}
			m, err := ParseMatcher(p)
			if err != nil {
				return nil, err
			}
			matchers = append(matchers, m)
		}
		if len(matchers) == 0 {
			return nil, fmt.Errorf("%w: compound pattern contains no valid patterns", MatchError)
		}
		if len(matchers) == 1 {
			return matchers[0], nil
		}
		return &CompoundMatcher{Matchers: matchers}, nil
	}

	// semicolon = OR (higher precedence)
	if strings.Contains(pattern, ";") {
		parts := strings.Split(pattern, ";")
		var matchers []Matcher
		for _, p := range parts {
			if strings.TrimSpace(p) == "" {
				continue
			}
			m, err := ParseMatcher(p)
			if err != nil {
				return nil, err
			}
			matchers = append(matchers, m)
		}
		if len(matchers) == 0 {
			return nil, fmt.Errorf("%w: union pattern contains no valid patterns", MatchError)
		}
		if len(matchers) == 1 {
			return matchers[0], nil
		}
		return &UnionMatcher{Matchers: matchers}, nil
	}

	return newFuzzyMatcher(pattern)
}
